### ISSUE TITLE
cannot reach the published performance

### ISSUE THREAD

**Issue by eastonYi** [espnet/espnet#642]
I run the HKUST and Librispeech recipes, but achieved the performances  far behind the published ones. I didn't use rnnlm since it costs much time. I wander whether it is normal to get there results since LM decoding cannot bring great improvement.

HKUST's  final validation log dring training:
```
    {
        "main/loss_ctc": 0.16775733435322762,
        "main/loss_att": 6.2305560690164565,
        "main/acc": 0.9195704427231433,
        "main/cer_ctc": 0.0012477977712993954,
        "main/cer": 0.0,
        "main/wer": 0.0,
        "main/loss": 3.1991566959023476,
        "validation/main/loss_ctc": 30.868107795715332,
        "validation/main/loss_att": 34.40174157278879,
        "validation/main/acc": 0.6124429075107217,
        "validation/main/cer_ctc": 0.4229814597370035,
        "validation/main/cer": 0.0,
        "validation/main/wer": 0.0,
        "validation/main/loss": 32.63492472512382,
        "eps": 1e-12,
        "epoch": 14,
        "iteration": 59300,
        "elapsed_time": 65910.05266291648
    },
```

HKUST's final results in stage 5:
```
+ echo 'write a CER (or TER) result in exp/train_nodup_sp_pytorch_vggblstm_e3_subsample1_2_2_1_1_unit1024_proj1024_d2_unit1024_location_aconvc10_aconvf100_mtlalpha0.5_adadelta_sampprob0.0_bs30_mli800_mlo150/decode_train_dev_beam1_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.6_rnnlm0.3_2layer_unit650_sgd_bs64/result.txt'
write a CER (or TER) result in exp/train_nodup_sp_pytorch_vggblstm_e3_subsample1_2_2_1_1_unit1024_proj1024_d2_unit1024_location_aconvc10_aconvf100_mtlalpha0.5_adadelta_sampprob0.0_bs30_mli800_mlo150/decode_train_dev_beam1_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.6_rnnlm0.3_2layer_unit650_sgd_bs64/result.txt
+ grep -e Avg -e SPKR -m 2 exp/train_nodup_sp_pytorch_vggblstm_e3_subsample1_2_2_1_1_unit1024_proj1024_d2_unit1024_location_aconvc10_aconvf100_mtlalpha0.5_adadelta_sampprob0.0_bs30_mli800_mlo150/decode_train_dev_beam1_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.6_rnnlm0.3_2layer_unit650_sgd_bs64/result.txt
|           SPKR                                        |           # Snt                      # Wrd           |           Corr                       Sub                        Del                        Ins                        Err                      S.Err           |
|           Sum/Avg                                     |           3999                       47130           |           69.8                      24.0                        6.1                        3.0                       33.2                       80.6           |
+ false
+ echo 'write a CER (or TER) result in exp/train_nodup_sp_pytorch_vggblstm_e3_subsample1_2_2_1_1_unit1024_proj1024_d2_unit1024_location_aconvc10_aconvf100_mtlalpha0.5_adadelta_sampprob0.0_bs30_mli800_mlo150/decode_dev_beam1_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.6_rnnlm0.3_2layer_unit650_sgd_bs64/result.txt'
write a CER (or TER) result in exp/train_nodup_sp_pytorch_vggblstm_e3_subsample1_2_2_1_1_unit1024_proj1024_d2_unit1024_location_aconvc10_aconvf100_mtlalpha0.5_adadelta_sampprob0.0_bs30_mli800_mlo150/decode_dev_beam1_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.6_rnnlm0.3_2layer_unit650_sgd_bs64/result.txt
+ grep -e Avg -e SPKR -m 2 exp/train_nodup_sp_pytorch_vggblstm_e3_subsample1_2_2_1_1_unit1024_proj1024_d2_unit1024_location_aconvc10_aconvf100_mtlalpha0.5_adadelta_sampprob0.0_bs30_mli800_mlo150/decode_dev_beam1_emodel.acc.best_p0.0_len0.0-0.0_ctcw0.6_rnnlm0.3_2layer_unit650_sgd_bs64/result.txt
|          SPKR                                        |          # Snt                      # Wrd           |          Corr                       Sub                       Del                        Ins                       Err                     S.Err           |
|          Sum/Avg                                     |          5413                       56154           |          71.3                      23.8                       5.0                        2.8                      31.5                      76.2           |
```

Librispeech's final validation log during training:
```
    {
        "main/loss_ctc": 22.70454153060913,
        "main/loss_att": 267.95261547088626,
        "main/acc": 0.13739110290527035,
        "main/cer_ctc": 0.05433470391174981,
        "main/cer": 0.0,
        "main/wer": 0.0,
        "main/loss": 145.32857885360718,
        "validation/main/loss_ctc": 27.41459705955104,
        "validation/main/loss_att": 172.43746476424369,
        "validation/main/acc": 0.13928295402855412,
        "validation/main/cer_ctc": 0.1256780020258113,
        "validation/main/cer": 0.0,
        "validation/main/wer": 0.0,
        "validation/main/loss": 99.92603005860981,
        "eps": 1e-14,
        "epoch": 13,
        "iteration": 34600,
        "elapsed_time": 67708.43240770698
    },
```

Besides, I need add `PYTHONIOENCODING=utf-8` before `./run.sh` to prevent getting unicode such as `prediction : \u2581AS\u2581I\u2581WAS\u2581TO\u2581SEE\u2581I...`. Adding this conmand in `run.sh` can not help. How to avoid it ? 

Thanks a lot !

**Comment by eastonYi** [comment]
By the way, I used pytorch 1.0.0 as the backend and build-in ctc for traing. Other settings are preserved as default.

**Comment by sw005320** [comment]
Thanks for the report. Could you run it with warp_ctc? I still found some strange behaviors in the build-in CTC. (The Librispeech attention loss is very high and this may not be due to the warp_ctc.) I just want to make sure this is a same setup. Also, did you use multiple GPUs? This also changes the performance.

**Comment by eastonYi** [comment]
Thanks for your replay! I used 4 gpus for training. I found error when using warper-ctc so I avoided using it. I put this in https://github.com/espnet/espnet/issues/647#issue-420852361

**Comment by stefbraun** [comment]
Also pytorch integrated CTC may or may not be deterministic on GPU(the algorithm may be chosen dynamically for each batch), while warp-ctc is deterministic on GPU (to the best of my knowledge).
See issue [17798](https://github.com/pytorch/pytorch/issues/17798)

**Comment by sw005320** [comment]
> I used 4 gpus for training

OK, this explains some of them. Multiple GPU corresponds to use larger minibatch, which often degrades the performance in the RNN training. If we use asynchronous update, we may avoid the issue, but we don't implement it yet. Also, again, I don't recommend to use the build-in CTC for now. In addition to the above deterministic issue, I feel like the current implementation does not correctly handle the long utterances based on the experience of my other experiments.

**Comment by vyouman** [comment]
@sw005320 So if I want to use a minibatch of N using n gpus, I should set the batchsize to be N/n then, right? I also found great performance degradation on WSJ_mix corpus using multi-gpu training with the default parameters.

**Comment by kan-bayashi** [comment]
@vyouman Right. If you use N gpus,batchsize will be automatically changed to N * (original batchsize).
Basically, our recipe can be performed using a single gpu with 12GB memory. 
So if you want to reproduce our results, we recommend to run the recipe with a single gpu.

And batch size have a larger effect for the performance especially in the case of small size task like wsj.
